File: hackerrank/30 days of code/day 28, regex, patterns, and intro to databases/regex_patterns_and_intro_to_databases.ts
// https://www.hackerrank.com/challenges/30-regex-patterns/problem?isFullScreen=true

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString: string = '';
let inputLines: string[] = [];
let currentLine: number = 0;

process.stdin.on('data', function (inputStdin: string): void {
  inputString += inputStdin;
});

process.stdin.on('end', function (): void {
  inputLines = inputString.split('\n');
  inputString = '';
  main();
});

//////////////////////////////////////////////////

function main() {
  const n: number = +readLine();
  const names: string[] = collectGmailUsers(n);
  names.sort();
  console.log(names.join('\n'));
}

function readLine(): string {
  return inputLines[currentLine++];
}

function collectGmailUsers(n: number): string[] {
  const names: string[] = [];

  while (n--) {
    const [name, email] = readLine().split(' ');
    if (email.endsWith('@gmail.com'))
      names.push(name);
  }

  return names;
}

File: hackerrank/30 days of code/day 28, regex, patterns, and intro to databases/regex_patterns_and_intro_to_databases.py
# https://www.hackerrank.com/challenges/30-regex-patterns/problem?isFullScreen=true

def main():
    n = int(input())
    names = collect_gmail_users(n)
    names.sort()
    print(*names, sep='\n')


def collect_gmail_users(n):
    names = []

    for _ in range(n):
        name, email = input().split()
        if email.endswith('@gmail.com'):
            names.append(name)

    return names


if __name__ == '__main__':
    main()
